### ISSUE TITLE
cabal clean etc leave build-breaking artifacts

### ISSUE THREAD

**Issue by TomMD** [GaloisInc/abcBridge#3]
I just had to dig from Verifier down to determine an old abcBridge 0.13 tarball downloaded by abcBridge setup was holding up the compilation of SAWScript.  We should probably make a more effective `cabal clean` if possible.

**Comment by atomb** [comment]
We do have `scripts/clean-abc.sh`, and it should be straightforward to update `Setup.hs` to run it during `cabal clean`.

**Comment by robdockins** [comment]
Tom: can you give more details on the error state you were experiencing?  At what point did the build break and what messages were you getting?

We could make `cabal clean` more aggressive, but I think it is a bad idea.  The same clean procedure should work for the released Hackage packages, which have the associated ABC sources bundled in; these released Hackage packages should never be fetching tarballs and such.  Maybe we could try to detect if we are building from git sources and have a more aggressive clean in that case?

I recognize the current situation is not ideal; I'm open to suggestions.

**Comment by TomMD** [comment]
@robdockins I've lost the state of the system that errored but the message was:

```
Setting up ABC tree for abcBridge version 0.14-dev...

The ABC source version 0.13-dev does not match the abcBridge package version 0.14-dev.

This may cause problems with the build; it is recommended you cancel this build and check out matching ABC sources
into the abc-build directory and try again.  You can run scripts/clean_abc.sh to remove the current ABC sources;
this will cause latest development branch to be automatically downloaded on the next build.

Press [enter] to continue anyway, or press [^C] to abort (build will automatically abort in 15 seconds).
```

Strictly speaking, the old tarball's existence wasn't the issue.  I believe manually (or otherwise) deleting `abc-build` was the key step to get the build system to download the correct version of ABC.  Sorry this is less than exact.

EDIT: I suppose my main point is that the above message is lost in the noise when building a higher-level library.  I had to dig through logs to come up with a message from a script that appeared to already know the exact action needed to obtain a functional build automatically.

**Comment by robdockins** [comment]
@TomMD: Indeed.  I've been reluctant to make the build system auto-delete the `abc-build` directory, but maybe that is a better default in the version mismatch case (which should never occur for released Hackage packages).

@acfolzer: We've had a discussion on this exact topic in the past.  Have you had any additional thoughts since then?

**Comment by TomMD** [comment]
@robdockins I understand the reluctance.  If this ticket falls into the NOTABUG or WONTFIX category I would understand.  The issue struck me as more significant earlier but I recognize that any regular user will acquire the code differently and without cruft of prior build artifacts.

**Comment by robdockins** [comment]
After discussion with @atomb and @acfoltzer, we decided to implement the following change: during the configure phase if we detect ABC sources that do not match the abcBridge cabal package version, we delete the abc-build directory and try to fetch sources.  A version mismatch should never occur for released packages, so that addresses my major concern.   With any luck, this should make the next version bump go more smoothly.

Fixed in 3359908a657cac032fb64b279a26a20d378f8c2d
